controller/scheduler-v2/scheduler.go

```diff
@@ -0,0 +1,1190 @@
+	return utils.JobConfig(req.Job.Formation.ExpandedFormation, req.Type, hostID)
+}
+
+func (s *Scheduler) findBestHost(formation *Formation, typ, hostID string) (utils.HostClient, error) {
+	log := fnLogger("app.id", formation.App.ID, "release.id", formation.Release.ID, "job.type", typ)
+	log.Info("getting host list")
+	hosts, err := s.getHosts()
+	if err != nil {
+		log.Error("error getting host list", "err", err)
+		return nil, err
+	}
+	if len(hosts) == 0 {
+		e := "no hosts found"
+		log.Error(e)
+		return nil, errors.New(e)
+	}
+
+	if hostID != "" {
+		log.Info("using explicit host", "host.id", hostID)
+		return s.Host(hostID)
+	}
+
+	counts := s.jobs.GetHostJobCounts(formation.key(), typ)
+	var minCount int = math.MaxInt32
+	for _, host := range hosts {
+		count, ok := counts[host.ID()]
+		if !ok || count < minCount {
+			minCount = count
+			hostID = host.ID()
+		}
+	}
+	log.Info(fmt.Sprintf("using host with least %s jobs", typ), "host.id", hostID)
+	return s.Host(hostID)
+}
+
+func (s *Scheduler) getHosts() ([]utils.HostClient, error) {
+	hosts, err := s.Hosts()
+
+	if err != nil {
+		return nil, err
+	}
+
+	// Ensure that we're only following hosts that we can discover
+	knownHosts := make(map[string]struct{})
+	for id, hostStream := range s.hostStreams {
+		if hostStream.Err() == nil {
+			knownHosts[id] = struct{}{}
+		} else {
+			s.unfollowHost(id)
+		}
+	}
+	for _, h := range hosts {
+		hostID := h.ID()
+		delete(knownHosts, hostID)
+		if _, ok := s.hostStreams[hostID]; !ok {
+			s.followHost(h)
+		}
+	}
+	for id := range knownHosts {
+		s.unfollowHost(id)
+	}
+	return hosts, nil
+}
+
+func (s *Scheduler) Stop() error {
+	log := fnLogger()
+	log.Info("stopping scheduler loop")
+	s.stopOnce.Do(func() {
+		close(s.stop)
+	})
+	return nil
+}
+
+func (s *Scheduler) Subscribe(events chan Event) stream.Stream {
+	log := fnLogger()
+	log.Info("adding event subscriber")
+	s.listenMtx.Lock()
+	defer s.listenMtx.Unlock()
+	s.listeners[events] = struct{}{}
+	return &Stream{s, events}
+}
+
+func (s *Scheduler) Unsubscribe(events chan Event) {
+	log := fnLogger()
+	log.Info("removing event subscriber")
+	s.listenMtx.Lock()
+	defer s.listenMtx.Unlock()
+	delete(s.listeners, events)
+}
+
+func (s *Scheduler) SaveJob(job *Job, appName string, status host.JobStatus, metadata map[string]string) bool {
+	switch status {
+	case host.StatusStarting:
+		s.handleJobEvent(job, JobStateStarting)
+	case host.StatusRunning:
+		s.handleJobEvent(job, JobStateRunning)
+	default:
+		if job.IsScheduled() {
+			s.handleJobCrash(job)
+		} else {
+			s.handleJobStop(job)
+		}
+	}
+	if !s.jobs.IsJobInState(job.JobID, job.state) {
+		log := fnLogger("job.id", job.JobID, "app.id", job.AppID, "app.name", appName, "release.id", job.ReleaseID, "job.type", job.Type, "job.status", status)
+		log.Info("queuing job for persistence")
+		s.putJobs <- controllerJobFromSchedulerJob(job, jobState(status), metadata)
+		return true
```

Comment:
It already did. This is a retroactive check. One of the reasons why copy-on-write is nice is that you can rely on `job` not mutating while handling it.